File: depth-first-search/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def height(self,root): 
        if root == None:
            return 0 
        return 1 + max(self.height(root.left), self.height(root.right)) 

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        
        # Check if subtrees have height difference within 1.
        # If they do, check if the subtrees are balanced
        return abs(self.height(root.left) - self.height(root.right)) < 2 and self.isBalanced(root.left) and self.isBalanced(root.right)

File: depth-first-search/balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean isBalanced = true; 

    public boolean isBalanced(TreeNode root) {
        getDepth(root);
        return this.isBalanced; 
    }

    int getDepth(TreeNode root){
        if (root == null){
            return 0; 
        }
        int leftDepth = getDepth(root.left);
        int rightDepth = getDepth(root.right); 
        if (Math.abs(rightDepth - leftDepth) > 1){
            this.isBalanced = false; 
        }
        return 1 + Math.max(leftDepth, rightDepth); 
    }
}